### ISSUE TITLE
Investigate slot security

### ISSUE THREAD

**Issue by JeroenDeDauw** [ProfessionalWiki/NeoWiki#350]
AC: Answer questions:
* Can we restrict READ access to slots? For instance, disallow access to the `neo` slot. 
* Does the web API allow for WRITE access to non-main slots? And if it does, can we restrict this?

https://www.mediawiki.org/wiki/API:Revisions

http://localhost:8484/index.php/Special:ApiSandbox#action=query&format=json&prop=revisions&titles=Professional_Wiki&rvprop=timestamp%7Cuser%7Ccomment%7Ccontent&rvslots=main%7Cneo

![image](https://github.com/ProfessionalWiki/NeoWiki/assets/146040/8737a446-0ec4-485b-b546-15dd8f20dde3)

**Comment by JeroenDeDauw** [comment]
Both questions answered below, against MediaWiki 1.43.6 source with the behaviour reproduced on a 1.43.8 dev wiki.

**Summary: there is no unguarded write path into the `neo` slot, and slot read cannot be restricted separately from page read because MediaWiki has no per-slot read primitive. Page read is necessarily the unit of access control.**

## Can we restrict READ access to slots?

**No**, and not by design choice — core offers no mechanism, on four independent axes:

- **`SlotRoleHandler`** exposes `getRole`, `getOutputLayoutHints`, `isDerived`, `getNameMessageKey`, `getDefaultModel`, `isAllowedModel` and `supportsArticleCount`. None takes an `Authority` or `User`. `isAllowedModel()` is a write-time content-model check, called only from `PageUpdater`.
- **`RevisionRecord`'s `$audience`** resolves to `audienceCan( DELETED_TEXT, … )` — the revision-wide RevDel bitfield, with no per-role dimension. `RevisionSlots::getContent()` takes no audience argument at all, and `RevisionRecord::getSlots()` is documented as "no audience checks applied".
- **No config setting.** The only slot-related one is `$wgRevisionSlotsCacheExpiry`, a cache TTL.
- **No hook.** The only slot-named hooks are `GetSlotDiffRenderer` and `TextSlotDiffRendererTablePrefix`, both diff rendering.

The slot is consequently readable on every core surface that exposes revision content, each gated on per-page `read`:

| Surface | Exposes `neo` |
|---|---|
| `action=query&prop=revisions&rvslots=neo` | yes |
| `action=raw&slot=neo` | yes — raw JSON served as `text/plain` |
| `Special:Export` | yes — as an MCR `<content>` element |
| `action=compare&slots=neo`, diff view | yes |
| `action=parse` | no — main slot only |
| Core REST `/v1/page`, `/v1/revision` | no — main slot only |

`action=raw&slot=neo` is the easiest to overlook: `SubjectContent` extends `JsonContent` extends `TextContent`, so it clears `RawAction`'s `instanceof TextContent` filter. The two "no" rows are API omissions, not permission gates.

This makes ADR 004's stated consequence ("only accessible via our own interfaces. This improves security and control") inaccurate — filed as #1166.

### One genuine leak, which is core's and not ours

Slot *metadata* escapes per-title read restrictions in `prop=revisions`:

- `rvprop=roles` emits `$revision->getSlotRoles()` with no read check at all (`ApiQueryRevisionsBase.php:421-423`), revealing that a page you may not read carries a `neo` slot.
- `rvslots=neo&rvprop=slotsize|slotsha1|contentmodel` returns the slot's byte size, SHA1 and content model. `needSlots` (`:312`) is satisfied by those three props without setting `fetchContent` (`:275`), and the pageset read gate is conditioned on `fetchContent` (`ApiQueryRevisions.php:196`). The per-revision `authorizeRead` at `:539` guards only the content value; `extractSlotInfo()` runs before it.

Calibrated: core-wide (the main slot leaks identically), metadata-only, and reachable only where per-page read restrictions are in play — though a SHA1 does allow confirming a guessed Subject JSON. Not fixable from an extension, and consistent with MediaWiki not guaranteeing per-page read restrictions across the API. Worth knowing for permission-filtered deployments; not tracked as a NeoWiki bug.

## Does the web API allow WRITE access to non-main slots, and can we restrict it?

**`action=edit` cannot touch a non-main slot.** `ApiEditPage` has no `slot`/`slots`/`slotrole` parameter and every `SlotRecord` reference in it is `::MAIN`; `EditPage` hardcodes the role on both the create and update branches. T200570 (MCR API editing) has not landed — no references anywhere in 1.43. Core REST's `EditHandler` is documented "main slot only" and proxies to `action=edit`. `action=changecontentmodel` is main-only.

Reproduced: `action=edit&…&slots=neo&slot=neo` returns `Unrecognized parameters: slots, slot` and writes the main slot only; forcing `contentmodel=NeoWikiSubject` is refused with `contentnotallowedhere`.

The paths that *can* write a non-main slot are all rights-gated:

| Path | Arbitrary content | Right |
|---|---|---|
| `action=import` / `Special:Import` | yes | `import` / `importupload` — sysop by default |
| `action=mcrundo`, `action=mcrrestore` | no — the page's own history | `edit` |
| `action=rollback` | no | `edit` + `rollback` |
| undelete / move / mergehistory | no — copy or inherit | respective rights |

Import is the only arbitrary-content writer, confirmed by importing XML that created a `neo` slot on a fresh page. Worth recording precisely, because it is weaker than the edit path: `ImportableOldRevisionImporter` bypasses `PageUpdater` entirely, building a `MutableRevisionRecord` and calling `RevisionStore::insertRevisionOn()` directly. So neither `PageUpdater::ensureRoleAllowed()` nor `SlotRoleHandler::isAllowedModel()` applies. Its only role check is `isDefinedRole()` (`ImportableOldRevisionImporter.php:148`), and when the XML supplies `<model>` it is used as given (`WikiImporter.php:1053-1056`) — so an importer can place an arbitrary content model into the `neo` slot. That is a data-integrity concern, not privilege escalation: it needs sysop rights, and a sysop can already do worse. Note that our `AfterImportPage` hook then projects imported slot content into Neo4j.

For the ordinary edit paths, `PageUpdater::ensureRoleAllowed()` rejects any role outside `SlotRoleRegistry`; core alone defines only `main`, so it is our own slot registration that makes `neo` a possible target at all.

NeoWiki's own write endpoints gate CSRF and then per-page `edit` (`AuthorityBasedSubjectAuthorizer` → `authorizeWrite( 'edit', $title )`). Verified live: anonymous `PUT /subject/{id}`, `POST /page/{id}/childSubjects` and `DELETE /subject/{id}` all answer `403`, at whichever of the two layers is reached first.

So "can we restrict this" is moot — it is already restricted, and there is nothing further to add.

## Disposition

Closing: both AC questions are answered and no unguarded path was found. Spun out:

- #1166 — correct ADR 004's inaccurate security claim.

Still open elsewhere and deliberately out of scope here: #87 (per-page permission checks) and #1059 (parser-time Cypher runs with no permission check), which is the live permission gap on the read side.

> <sub>AI-authored — Claude Code, `Fable 5 (max)`; one-line ask from @JeroenDeDauw, no revisions; not yet human-reviewed; claims checked against MediaWiki 1.43.6 source and reproduced on a 1.43.8 dev wiki.</sub>
